rfc-142-split-database-instances.md

```diff
@@ -0,0 +1,50 @@
+# One database per RDS instance
+
+## Summary
+
+We currently use a one-instance-many-databases approach to hosting MySQL, Postgres, Mongo and Redis. These mega-databases make it difficult to perform major version upgrades – they need to be coordinated across all applications at the same time, increase the blast-radius of problems in a single database, and make it harder to appropriately size and monitor resources for an application.
+
+This RFC proposes that we split the single RDS instances for MySQL and Postgres, moving to a single RDS instance per database. It also proposes we apply the same principle to our self-hosted Mongo and Redis installations.
+
+This principle _could_ be applied to RabbitMQ, S3, etc., but it's out of scope for this RFC.
+
+## Naming
+
+The word "database" is often interchangeably used to describe a collection of tables, and an instance of (e.g.) Postgres server running somewhere. For the purposes of this RFC, "database" refers to the collection of tables, schema, etc.. We'll use "instance" to refer to the runtime that hosts the database.
+
+## Problem
+
+Having a single instance hosting multiple databases has a number of drawbacks:
+
+1. Major upgrades must be applied to all databases in one go, as it's the instance we're upgrading, not the database. Major upgrades are considered risky because they contain potentially breaking, and backward-incompatible changes. This means major upgrades are scarier, so we're less likely to do them until we're forced. They also require coordination across disparate teams so everyone's confident their apps will survive the upgrade.
+
+2. Problems in one database are likely to impact other unrelated services. For example an app with unexpectedly high load on its database could max-out the instance, causing performance degradation in other databases and their apps.
+
+3. Related to [2], it's significantly harder to size instance resources. Disk is fairly straightforward, but it's harder to understand CPU/RAM/IOPS patterns. We're also likely to over-size resources to reduce the likelihood large-scale problems in [2].
+
+4. There is no clear owner of the instances. Each app is owned by a single team, and each database should be accessed by exactly one app. However many apps with different owners reside on the single instance. Sizing and managing databases is an app-level responsibility, so it doesn't fit with the remit of Platform Reliability/Replatforming/Platform Health. Having multiple databases on a single instance 
+
+This is coming up now because our versions of MySQL and Postgres are approaching End of Life (notice for [MySQL](https://forums.aws.amazon.com/ann.jspa?annID=8790) and [Postgres](https://forums.aws.amazon.com/ann.jspa?annID=8499)), forcing us to plan major upgrades for both over the next few months.
+
+## Proposal
+
+We propose that:
+
+1. All new long-lived databases are created in their own individual RDS instances.
```

Comment:
Could we make this slightly more vague and just refer to `instances` rather than `RDS instances`, to allow individual teams to explore non-RDS options should they wish to?

Reply:
That's a good question @brucebolt. Do you have something in mind, or is this just a concern that we're being overly prescriptive here?

I'm not tied to "RDS" specifically, but there could be a readability problem as we make it more generic. For example, it's easy enough to grok "Split these three MySQL databases into 3 Aurora instances", but as we broaden out to (potentially) all supporting services we're reduced to the RFC saying "distinct supporting services in separate instances". "Instances" has a really specific meaning to me now I've looked at this RFC for a while, but I wonder if it's intuitive.

I'd be tempted to err on the side of specificity if we consistently use AWS services at the moment, with the caveat that teams should always feel able to explore other options if the "standard" way doesn't meet their needs... but if things aren't that consistent then :+1: to removing this. 

📣 @kevindew this ties to your point about broadening the RFC out

Reply:
I think my concern is just about this being too prescriptive.  In future, teams may look back at this and think we'd specified RDS for a reason that we didn't note, but could've been important.  But I'm happy for RDS to remain specified, as that's what we're currently using.

Reply:
Is the general point here to move away from shared persistent storage, whether it be an RDS instance, DBs on VMs or S3 buckets? So should the wording here be something more like "persistent storage should not be shared across application boundaries, e.g. separate RDS instances, S3 buckets, elasticache instances, etc"?

Reply:
I think that's a valid point too @kerin. Here I'm referring to long-lived compute as well as storage – the wording is a bit awkward but I was alluding to CI, where we might have long-lived instances which house short-lived databases that are created & destroyed in a test cycle.